src/rapids_pre_commit_hooks/copyright.py

```diff
@@ -0,0 +1,230 @@
+                    if (
+                        int(match.group("last_year") or match.group("first_year"))
+                        < current_year
+                    ):
+                        linter.add_warning(
+                            match.span("years"), "copyright is out of date"
+                        ).add_replacement(
+                            match.span(),
+                            f"Copyright (c) {match.group('first_year')}-{current_year}"
+                            ", NVIDIA CORPORATION",
+                        )
+            else:
+                linter.add_warning((0, 0), "no copyright notice found")
+
+
+def get_target_branch(repo):
+    # Try environment
+    target_branch_name = os.getenv("GITHUB_BASE_REF")
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+    target_branch_name = os.getenv("TARGET_BRANCH")
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+    target_branch_name = os.getenv("RAPIDS_BASE_BRANCH")
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+
+    # Try config
+    with repo.config_reader() as r:
+        target_branch_name = r.get("rapidsai", "baseBranch", fallback=None)
+    if target_branch_name:
+        try:
+            return repo.heads[target_branch_name]
+        except IndexError:
+            pass
+
+    # Try newest branch-xx.yy
+    branches = sorted(
+        (
+            (branch, (match.group("major"), match.group("minor")))
+            for branch in repo.heads
+            if (match := BRANCH_RE.search(branch.name))
+        ),
+        key=lambda i: i[1],
+        reverse=True,
+    )
+    try:
+        return branches[0][0]
+    except IndexError:
+        pass
+
+    # Appropriate branch not found
+    return None
+
+
+def get_target_branch_upstream_commit(repo):
+    target_branch = get_target_branch(repo)
+    if target_branch is None:
+        try:
+            return repo.head.commit
+        except ValueError:
+            return None
+
+    target_branch_upstream = target_branch.tracking_branch()
+    if target_branch_upstream:
+        return target_branch_upstream.commit
+
+    def try_get_ref(remote):
+        try:
+            return remote.refs[target_branch.name]
+        except IndexError:
+            return None
+
+    candidate_upstreams = sorted(
+        (upstream for remote in repo.remotes if (upstream := try_get_ref(remote))),
+        key=lambda upstream: upstream.commit.committed_datetime,
+        reverse=True,
+    )
+    try:
+        return candidate_upstreams[0].commit
+    except IndexError:
+        pass
+
+    return target_branch.commit
+
+
+def get_changed_files():
+    try:
+        repo = git.Repo()
+    except git.InvalidGitRepositoryError:
+        return {
```

Comment:
Does it make sense for a pre-commit hook to be runnable without a git repo? I believe pre-commit requires git for its basic operations. Just wanting to avoid engineering/testing for a case that we can't reach.

Reply:
Tools like `clang-format` can run just fine without a Git repo, and I think this tool should as well. Having Git history available tells us whether or not the file has changed in a way that requires a copyright update, but if it's not in a repo, we can fall back to always updating the copyright. The cost of supporting this use case is low IMO.

Reply:
Perfectly fine -- you're right, this is better to design as a standalone tool that happens to provide pre-commit configurations. And it does seem the cost is low.